src/index.ts

```diff
@@ -0,0 +1,229 @@
+/*
+ * A stack-machine based Example syntax parser.
+ *
+ * Spec:
+ * 
+ *     A text is considered a named fragment when it is located between
+ *     "// BEGIN FRAGMENT: <name>" and "// END FRAGMENT", <> not included.
+ *     This syntax is case-sensitive.
+ * 
+ *     Don't use special characters in names. Prefer snake case.
+ * 
+ *     Fragments can be included in one another.
+ *     In that case the lines matching
+ *     "// BEGIN FRAGMENT: " and "// END FRAGMENT" are removed.
+ *     Elements between "// BEGIN ESCAPE" and "// END ESCAPE"
+ *     are excluded unconditionally from the tutorial.
+ * 
+ * Example code:
+ * 
+ *     import { ExampleParser } from "./example_parser"
+ *     let text: string = "// BEGIN FRAGMENT: test\nHello.\n// END FRAGMENT"
+ *     let parser_inst: ExampleParser = new ExampleParser(text)
+ *     console.log(parser_inst.mapLines())
+ * 
+ */
+
+// A custom error type to be used with broken syntax
+export class ExampleSyntaxError extends Error {
+    constructor(message:string = '') {
+      super(message)
+      this.message = message
+    }
+}
+
+// Obtain a range of numbers as an array, including the end
+let range = function(start: number, end: number): Array<number> {
+    let src: Array<number> = Array.from(Array(end - start + 1).keys())
+    return src.map(x => x + start)
+}
+
+// Check if the string is empty or contains
+// only the space and tabulation characters
+function emptySpace(str: string) {
+    return str.trim().length === 0
+}
+
+// Remove empty strings at both start and end
+// in an array of lines
+export function trimEmptyLines(lines: Array<string>) {
+    // Remove the first and the last string if those are
+    // filled with spaces or empty
+    while (emptySpace(lines.at(0) || '')) lines.shift()
+    while (emptySpace(lines.at(-1) || '')) lines.pop()
+    return lines
+}
+
+// Clear line indentation for an array of strings containing the lines
+let clearIndentation = function(lines: Array<string>) {
+    // Calculate a padding common for all the lines
+    let commonIndent: number = getCommonStartPadding(lines)
+    // Return the lines with a common indentation removed
+    return lines.map(
+        line => {return line.slice(commonIndent) }
+    )
+}
+
+// Checks if whitespaces and tabs of a given prefix length are equal for all provided lines
+function spaceSectionEquals(lines: Array<string>, prefixLen: number) {
+    let prefixes = lines.map(line => {return line.slice(0, prefixLen)})
```

Comment:
```suggestion
    let prefixes = lines.map(line => line.slice(0, prefixLen))
```

Reply:
Diff got outdated, so I've updated this part separately, thanks!